File: Data Structures & Algorithms/decode-ways/submission-3.cpp
class Solution {
public:
int solve(int index,string s,vector<int>& dp){
    if(index==s.size()) return 1;
    if(s[index]=='0') return 0;
    if(dp[index]!=-1) return dp[index];
    int res=solve(index+1,s,dp);
    if(index<s.size()-1){
        if(s[index]=='1' || s[index]=='2' && s[index+1]<'7')
        res+=solve(index+2,s,dp);
    }
    return dp[index]=res;
}
    int numDecodings(string s) {
        vector<int> dp(s.size(),-1);
        return solve(0,s,dp);
    }
};

File: Data Structures & Algorithms/coin-change/submission-3.java
class Solution {

    int solve(int index, int[] coins, int amount, int[][] dp) {

        if (amount == 0)
            return 0;

        if (amount < 0 || index == coins.length)
            return 1_000_000_000;

        if (dp[index][amount] != -1)
            return dp[index][amount];

        int take = 1_000_000_000;

        if (coins[index] <= amount) {
            take = 1 + solve(index, coins, amount - coins[index], dp);
        }

        int notTake = solve(index + 1, coins, amount, dp);

        return dp[index][amount] = Math.min(take, notTake);
    }

    public int coinChange(int[] coins, int amount) {

        int[][] dp = new int[coins.length][amount + 1];

        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }

        int ans = solve(0, coins, amount, dp);

        return ans == 1_000_000_000 ? -1 : ans;
    }
}